examples/server/server.cpp

```diff
@@ -1,46 +1,116 @@
 
 struct server_params
 {
   std::string hostname = "127.0.0.1";
   int32_t port = 8080;
+  int32_t read_timeout = 600;
+  int32_t write_timeout = 600;
+  bool verbose = false;
+};
+
+static size_t common_part(const std::vector<llama_token> & a, const std::vector<llama_token> & b) {
+  size_t i;
+  for (i = 0; i < a.size() && i < b.size() && a[i] == b[i]; i++);
+  return i;
+}
+
+enum stop_type {
+    STOP_FULL,
+    STOP_PARTIAL,
 };
 
+bool ends_with(const std::string &str, const std::string &suffix)
+{
+    return str.size() >= suffix.size() &&
+           0 == str.compare(str.size() - suffix.size(), suffix.size(), suffix);
+}
+
+size_t find_partial_stop_string(const std::string &stop, const std::string &text)
+{
+    if (!text.empty()) {
+        const char text_last_char = text.back();
+        for (int64_t char_index = stop.size() - 1; char_index >= 0; char_index--) {
+            if (stop[char_index] == text_last_char) {
+                const std::string current_partial = stop.substr(0, char_index + 1);
+                if (ends_with(text, current_partial)) {
+                    return text.size() - char_index - 1;
+                }
+            }
+        }
+    }
+    return std::string::npos;
+}
+
+static std::string debug_str(const std::string & s) {
+  std::string ret;
+  for (size_t i = 0; s[i]; i++) {
+    switch (s[i]) {
+      case '\n': ret += "\\n"; break;
+      case '"': ret += "\\\""; break;
+      default: ret += s[i]; break;
+    }
+  }
+  return ret;
+}
+
+template<class InputIt, class OutputIt>
+static std::string tokens_to_str(llama_context * ctx, InputIt begin, OutputIt end) {
+  std::string ret;
+  for (; begin != end; (void)++begin) {
+    ret += llama_token_to_str(ctx, *begin);
+  }
+  return ret;
+}
+
 struct llama_server_context
 {
-  bool as_loop = false;
+  bool stream = false;
   bool has_next_token = false;
   std::string generated_text = "";
 
-  int32_t num_tokens_predicted = 0;
-  int32_t n_past = 0;
-  int32_t n_consumed = 0;
-  int32_t n_session_consumed = 0;
-  int32_t n_remain = 0;
+  size_t num_tokens_predicted = 0;
+  size_t n_past = 0;
+  size_t n_remain = 0;
 
   std::vector<llama_token> embd;
   std::vector<llama_token> last_n_tokens;
-  std::vector<llama_token> processed_tokens;
-  std::vector<llama_token> llama_token_newline;
-  std::vector<llama_token> embd_inp;
-  std::vector<std::vector<llama_token>> no_show_words;
-  std::vector<llama_token> tokens_predicted;
 
-  llama_context *ctx;
+  llama_context *ctx = nullptr;
   gpt_params params;
 
+  std::string stopping_word;
+
+  bool verbose = false;
+  int json_indent = -1;
+  int32_t multibyte_pending = 0;
+
+  ~llama_server_context()
+  {
+      if (ctx) {
+          llama_free(ctx);
+          ctx = nullptr;
+      }
+  }
+
   void rewind() {
-    as_loop = false;
     params.antiprompt.clear();
-    no_show_words.clear();
     num_tokens_predicted = 0;
     generated_text = "";
+    generated_text.reserve(params.n_ctx);
```

Comment:
It can expand a lot more beyond one context full of text because of truncation. You could use `n_predict` instead, but this can be turned off as well and then it's pretty much unlimited.

Reply:
let's leave it like that. sounds like the least bad workarounds :laughing: